src/Evolution/Systems/Cce/PrecomputeCceDependencies.hpp

```diff
@@ -0,0 +1,261 @@
+#include "Evolution/Systems/Cce/Tags.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "NumericalAlgorithms/Spectral/SwshCollocation.hpp"
+#include "NumericalAlgorithms/Spectral/SwshTags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/VectorAlgebra.hpp"
+
+namespace Cce {
+
+namespace detail {
+// A convenience function for computing the spin-weighted derivatives of \f$R\f$
+// divided by \f$R\f$, which appears often in Jacobians to transform between
+// Bondi coordinates and the numerical coordinates used in CCE.
+template <typename DerivKind>
+void angular_derivative_of_r_divided_by_r_impl(
+    gsl::not_null<
+        SpinWeighted<ComplexDataVector,
+                     Spectral::Swsh::Tags::derivative_spin_weight<DerivKind>>*>
+        d_r_divided_by_r,
+    const SpinWeighted<ComplexDataVector, 0>& boundary_r, size_t l_max,
+    size_t number_of_radial_points) noexcept;
+
+}  // namespace detail
+
+/*!
+ * \brief A set of procedures for computing the set of inputs to the CCE
+ * integrand computations that can be computed before any of the intermediate
+ * integrands are evaluated.
+ *
+ * \details The template specializations of this template are
+ * compatible with acting as a the mutator in a \ref DataBoxGroup
+ * `db::mutate_apply` operation. For flexibility in defining the \ref
+ * DataBoxGroup structure, the tags for `Tensor`s used in these functions are
+ * also organized into type lists:
+ * -  type alias `integration_independent_tags`: with a subset of
+ * `Cce::pre_computation_tags`, used for both input and output.
+ * -  type alias `boundary_values`: with a subset of
+ *   `Cce::pre_computation_boundary_tags`, used only for input.
+ * - type alias `pre_swsh_derivatives` containing hypersurface quantities. For
+ * this struct, it will only ever contain `Cce::Tags::BondiJ`, and is used as
+ * input.
+ *
+ * The `BoundaryPrefix` tag allows easy switching between the
+ * regularity-preserving version and standard CCE
+ *
+ */
+template <template <typename> class BoundaryPrefix, typename Tag>
+struct PrecomputeCceDependencies;
+
+/// Computes \f$1 - y\f$ for the Cce system.
+template <template <typename> class BoundaryPrefix>
+struct PrecomputeCceDependencies<BoundaryPrefix, Tags::OneMinusY> {
+  using boundary_tags = tmpl::list<>;
+  using pre_swsh_derivative_tags = tmpl::list<>;
+  using integration_independent_tags = tmpl::list<>;
+
+  using return_tags = tmpl::list<Tags::OneMinusY>;
+  using argument_tags = tmpl::list<Spectral::Swsh::Tags::LMax,
+                                   Spectral::Swsh::Tags::NumberOfRadialPoints>;
+
+  static void apply(
+      const gsl::not_null<Scalar<SpinWeighted<ComplexDataVector, 0>>*>
+          one_minus_y,
+      const size_t l_max, const size_t number_of_radial_points) noexcept {
+    const size_t number_of_angular_points =
+        Spectral::Swsh::number_of_swsh_collocation_points(l_max);
+    const auto& one_minus_y_collocation =
```

Comment:
I think you want this to be `DataVector` instead of `auto&` so that you don't end up with an expression template object and redo the subtraction for each radial point